### PR TITLE
WITH ORDER support on CREATE EXTERNAL TABLE

### LINKED ISSUE #5372: Support providing order information on CREATE EXTERNAL TABLE
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I want to be able to read an external file with sort information. 

```sql
CREATE EXTERNAL TABLE t(c1 int, ts int) STORED AS CSV ORDERED BY (ts) LOCATION 'foo.csv'
```

This query will enhance the physical planner by providing sort information.

We need to add this information `output_ordering` in `FileScanConfig`like `table_partition_cols`. `ListingOptions` is enough to support this feature.

**Describe the solution you'd like**
Enhance the parser and adapt the tests. Then, feed this information into `FileScanConfig`.  

**Describe alternatives you've considered**
NA

**Additional context**
NA

### PR DESCRIPTION
**Pull request by metesynnada**
# Which issue does this PR close?

Closes #5372.

# Rationale for this change

When working with files, specifying column orders can be highly effective. By leveraging pre-sorted files, we can avoid the need to materialize and sort the entire dataset, resulting in significant performance gains.

# What changes are included in this PR?

- `WITH ORDER` support on the parser.
- Necessary adjustments on `CreateExternalTable` (parser, logical plan, proto).
- `ListingTable` order support from SQL queries

# Are these changes tested?

Yes

# Are there any user-facing changes?

Now,

```sql
CREATE EXTERNAL TABLE t(c1 int) STORED AS CSV WITH ORDER (c1 ASC) LOCATION 'foo.csv'
```

is supported.

### CODE REVIEW

File: datafusion/sql/src/statement.rs [hunk 1]
```diff
@@ -425,6 +425,34 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        schema: &DFSchemaRef,
+    ) -> Result<Vec<datafusion_expr::Expr>> {
+        // Convert each OrderByExpr to a SortExpr:
+        let result = order_exprs
+            .into_iter()
+            .map(|e| self.order_by_to_sort_expr(e, schema))
+            .collect::<Result<Vec<_>>>()?;
+
+        // Verify that columns of all SortExprs exist in the schema:
+        for expr in result.iter() {
```
Comment (comphead) [inline datafusion/sql/src/statement.rs:440]: nit: do we really need to loop twice here?

**Review by comphead — COMMENTED** [review]
Thanks @metesynnada for the PR, looks great

File: datafusion/sql/src/parser.rs [hunk 2]
```diff
@@ -713,6 +782,79 @@ mod tests {
         });
         expect_parse_ok(sql, expected)?;
 
+        // Ordered Col
+        let sql = "CREATE EXTERNAL TABLE t(c1 int, c2 int) STORED AS CSV WITH ORDER (c1 ASC, c2 DESC NULLS FIRST) LOCATION 'foo.csv'";
```
[the remark this answers is not in the document]
Comment (metesynnada) [reply datafusion/sql/src/parser.rs:786]: I can add tests, ofc.

File: datafusion/sql/src/parser.rs [hunk 3]
```diff
@@ -253,6 +256,49 @@ impl<'a> DFParser<'a> {
...
+    /// Parse an ORDER BY sub-expression optionally followed by ASC or DESC.
+    pub fn parse_order_by_expr(&mut self) -> Result<OrderByExpr, ParserError> {
+        let expr = self.parser.parse_expr()?;
+
+        let asc = if self.parser.parse_keyword(Keyword::ASC) {
+            Some(true)
+        } else if self.parser.parse_keyword(Keyword::DESC) {
+            Some(false)
+        } else {
+            None
```
[the remark this answers is not in the document]
Comment (metesynnada) [reply datafusion/sql/src/parser.rs:281]: It was the default, I so kept it. It means there is no order, identical to `vec![]`.

File: datafusion/sql/src/statement.rs (hunk 1 above)
Comment (comphead) [reply datafusion/sql/src/statement.rs:440]: Right, I was thinking about we can iterate just once and return resuls if everything is ok or error if something is wrong https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect

**Comment by metesynnada** [comment]
@alamb I think you have use cases for parquet, does this enough to cover yours as well?

File: datafusion/core/tests/sql/order.rs [hunk 4]
```diff
@@ -39,3 +42,31 @@ async fn sort_with_lots_of_repetition_values() -> Result<()> {
...
+    let ctx = SessionContext::new();
+    let sql = "CREATE EXTERNAL TABLE dt (a_id integer, a_str string, a_bool boolean) STORED AS CSV WITH ORDER (a_id ASC) LOCATION 'file://path/to/table';";
+    if let LogicalPlan::CreateExternalTable(cmd) =
+        ctx.state().create_logical_plan(sql).await?
+    {
+        let listing_table_factory = Arc::new(ListingTableFactory::new());
+        let table_dyn = listing_table_factory.create(&ctx.state(), &cmd).await?;
+        let table = table_dyn.as_any().downcast_ref::<ListingTable>().unwrap();
+        assert_eq!(
+            &cmd.order_exprs,
```
Comment (alamb) [inline datafusion/core/tests/sql/order.rs:57]: I would recommend explicitly asserting the content of the order by exprs here -- as written if `cmd.order_exprs` were empty the test would still pass, I think.

File: datafusion/sql/src/parser.rs [hunk 5]
```diff
@@ -713,11 +782,144 @@ mod tests {
...
+                    asc: Some(true),
+                    nulls_first: None,
+                },
+                OrderByExpr {
+                    expr: Identifier(Ident {
+                        value: "c2".to_owned(),
+                        quote_style: None,
+                    }),
+                    asc: Some(false),
+                    nulls_first: Some(true),
```
Comment (alamb) [inline datafusion/sql/src/parser.rs:864]: I didn't see any test for `nulls_first: Some(false)`

File: datafusion/expr/src/logical_plan/plan.rs [hunk 6]
```diff
@@ -1632,6 +1632,8 @@ pub struct CreateExternalTable {
     pub if_not_exists: bool,
     /// SQL used to create the table, if available
     pub definition: Option<String>,
+    /// Order expressions
```
Comment (alamb) [inline datafusion/expr/src/logical_plan/plan.rs:1635]:
```suggestion
    /// Order expressions supplied by user.
    ///
    /// NOTE: DataFusion will assume the data is sorted in this way
    /// and does not verify it. Thus if the data is not actually sorted as
    /// described, WRONG ANSERS can be produced. 
```

**Review by alamb — APPROVED** [review]
This looks great to me -- thank you @metesynnada 

I had some suggestions on additional tests, 

My biggest concern is to make sure we document the implications of this ORDER BY sufficiently. Specifically, if the user creates a table like `CREATE EXTERNAL TABLE ... ORDER BY (a)` and the data is not actually sorted by `a` wrong results can occur I think

In addition to the note I suggested in `CreateExternalTable` can you please also add a description in the user guide: 

https://github.com/apache/arrow-datafusion/blob/main/docs/source/user-guide/sql/ddl.md#create-external-table

**Comment by metesynnada** [comment]
@alamb I think this is ready to merge unless you have additional suggestions on the document.

File: docs/source/user-guide/sql/ddl.md [hunk 7]
```diff
@@ -98,6 +98,12 @@ WITH ORDER (sort_expression1 [ASC | DESC] [NULLS { FIRST | LAST }]
          [, sort_expression2 [ASC | DESC] [NULLS { FIRST | LAST }] ...])
 ```
 
+#### Cautions When Using the WITH ORDER Clause
+
+- It's important to understand that using the `WITH ORDER` clause in the `CREATE EXTERNAL TABLE` statement only specifies the order in which the data should be read from the external file. If the data in the file is not already sorted according to the specified order, then the results may not be correct.
```
Comment (alamb) [inline docs/source/user-guide/sql/ddl.md:103]: 👍

**Review by alamb — APPROVED** [review]
Looks great -- thank you @metesynnada  and @comphead

**Comment by alamb** [comment]
I'll plan to merge when CI is clean

**Pull request merged**